src/commands/dev.ts

```diff
@@ -175,6 +175,12 @@ let currentlyRunningCommand: any = null;
 export async function command(commandOptions: CommandOptions) {
   const {cwd, config} = commandOptions;
   const {port: defaultPort, hostname, open, hmr: isHmr} = config.devOptions;
+  // WORKAROUND: Dev always assumes that you're serving from the root URL of your dev server.
```

Comment:
Here's the workaround. I think it makes sense to attach a private "dev mode" boolean to the config object so that our shared build logic can react to things like this.

Alternatively, we can pass an isDev property down as an argument to our individual `wrap*`  functions, and anywhere else that we resolve these URLs, similar to how we pass `hasHmr`.

@drwpow thoughts?

Reply:
Overall I think this is a needed change we should do now. But I’m -1 in general for storing state like this on the config object.

I’m wondering if we could instead take the 3 major operations of Snowpack—Install, Build, Dev—and instead refactor them a bit to be more object-oriented in a way that saves state, e.g.:

```
const Install = new Installer(config);

class Install {
  constructor(config) {
    // parse config and save stuff in state here
  }
}
```

I’ve noticed in some instances we go beyond merely normalizing config. [This line in particular](https://github.com/pikapkg/snowpack/blob/master/src/config.ts#L560) stores live plugins in the config and blows away the user’s original settings. We also have underscored “private” state objects stored in the config, too, as a temporary measure.

To be clear: I don’t think it’s posed a big problem up till now, but I think we do need to keep track of more and more state, and we do so within config. I think having a better place to store things like instantiated plugins, environment-specific settings, etc. would make the code easier to reason, and probably clean up a lot of the functions, too.

Reply:
+1, agreed with everything said above in principle. We can also look into having a `context` object that stores "state" (different from config) like this. Something like:

```
new Installer(config[, context]);
```

Or, maybe state is something internal to each command, in which case it could be managed internally. Not sure if functional vs. OOP really changes anything too much, although I definitely get the appeal for the dev server, which is stateful by definition.